google/cloud/forseti/scanner/audit/enabled_apis_rules_engine.py

```diff
@@ -0,0 +1,401 @@
+# Copyright 2018 The Forseti Security Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Enabled APIS rules engine for organizations, folders, and projects.
+
+Builds the RuleBook (EnabledApisRuleBook) from the rule definitions (file either
+stored locally or in GCS) and compares a list of APIs  against the RuleBook to
+determine whether there are violations.
+"""
+
+import collections
+import itertools
+import threading
+
+from google.cloud.forseti.common.gcp_type import errors as resource_errors
+from google.cloud.forseti.common.gcp_type import resource as resource_mod
+from google.cloud.forseti.common.gcp_type import resource_util
+from google.cloud.forseti.common.util import logger
+from google.cloud.forseti.common.util import relationship
+from google.cloud.forseti.scanner.audit import base_rules_engine as bre
+from google.cloud.forseti.scanner.audit import errors as audit_errors
+
+LOGGER = logger.get_logger(__name__)
+
+# Rule Modes.
+_WHITELIST = 'whitelist'
+_BLACKLIST = 'blacklist'
+_REQUIRED = 'required'
+_RULE_MODES = frozenset([_WHITELIST, _BLACKLIST, _REQUIRED])
+
+# Rule violation types.
+_VIOLATION_TYPES = {
+    _WHITELIST: 'ENABLED_APIS_ADDED_VIOLATION',
+    _BLACKLIST: 'ENABLED_APIS_ADDED_VIOLATION',
+    _REQUIRED: 'ENABLED_APIS_REMOVED_VIOLATION',
+}
+
+
+def _check_whitelist_apis(rule_apis=None, enabled_apis=None):
+    """Whitelist: Checks that enabled APIs ARE in rule APIs.
+
+    If an enabled api is NOT found in the rule APIs, add it to
+    the violating APIs.
+
+    Args:
+        rule_apis (list): ManagedService APIs allowed in the rule.
+        enabled_apis (list): ManagedService APIs enabled by the project.
+
+    Returns:
+        list: Enabled APIs NOT found in the whitelist (rule APIs).
+    """
+    return [api for api in enabled_apis if api not in rule_apis]
```

Comment:
We made a decision recently that this style of code is really hard to troubleshoot, because our IDE debuggers can not step through it, and so should not be used anymore.  So, can you please expand this out a bit more?

Can you please do this everywhere?
```
violating_apis = []
for api in enabled_apis:
    if api not in rule_apis:
        violating_apis.append(api)
return violating_apis

Reply:
I think I have changed them all.